src/shims/unix/fd.rs

```diff
@@ -180,14 +197,51 @@ impl FileDescriptor {
         RefMut::map(self.0.borrow_mut(), |fd| fd.as_mut())
     }
 
-    pub fn close<'ctx>(self, communicate_allowed: bool) -> InterpResult<'ctx, io::Result<()>> {
+    pub fn close<'tcx>(
+        self,
+        ecx: &mut MiriInterpCx<'tcx>,
+        communicate_allowed: bool,
+    ) -> InterpResult<'tcx, io::Result<()>> {
         // Destroy this `Rc` using `into_inner` so we can call `close` instead of
         // implicitly running the destructor of the file description.
         match Rc::into_inner(self.0) {
-            Some(fd) => RefCell::into_inner(fd).close(communicate_allowed),
+            Some(fd) => RefCell::into_inner(fd).close(ecx, communicate_allowed),
             None => Ok(Ok(())),
         }
     }
+    pub fn downgrade(&self) -> WeakFileDescriptor {
+        WeakFileDescriptor(Rc::downgrade(&self.0))
+    }
+}
+
+// WeakFileDescriptor is used in epoll ready_list and interest_list to avoid strong references,
+// so the file description can be closed properly.
+#[derive(Clone, Debug, Default)]
+pub struct WeakFileDescriptor(Weak<RefCell<Box<dyn FileDescription>>>);
+
+impl WeakFileDescriptor {
+    pub fn upgrade(&self) -> Option<Rc<RefCell<Box<dyn FileDescription>>>> {
+        self.0.upgrade()
+    }
+}
+
+impl PartialOrd for WeakFileDescriptor {
```

Comment:
The order of `impl` here is very confusing, I had to read this twice to make any sense of it.^^
Please use the natural order of the traits: PartialEq, Eq, PartialOrd, Ord.

Also -- having FDs be ordered is a surprise. Please explain what the order actually is.

Reply:
These four traits is implemented because it is required by ``BTreeMap`` and the actual order actually doesn't matter except that it will cause the execution to be non-deterministic.